### ISSUE TITLE
db: excessive compaction prioritization on drt-chaos

### ISSUE THREAD

**Issue by itsbilal** [cockroachdb/pebble#4052]
When investigating cockroachdb/cockroach#132668, I noticed excessively large tombstone density compactions on drt-chaos. These compactions involved way more files than the avg default compaction:

```
[n2,s8,pebble] 309152  [JOB 20240] compacted(tombstone-density) L5 [6673464 6673456 6673808 6673642] (684KB) Score=0.00 + L6 [6670570 6668014 6668644 6668649 6673055 6673060 6673056 6672933 6672929 6667641 6672760 6668752 6668756 6668753 6668426 6670681 6670688 6670682 6669574 6669571 6669070 6672176 6672181 6673403 6673471 6673404 6668900 6670804 6670811 6670805 6664304 6665940 6665945 6669620 6669639 6672344 6672349 6672345 6669103 6669098 6667202 6673599 6673836 6673600 6671869 6671863 6671832 6671828 6667445 6667440 6664316 6673509 6673590 6673510 6669588 6669585 6668044 6670713 6670721 6673769 6673761 6673927 6673919 6673928 6673776 6673785 6673694 6673703 6673991] (85MB) Score=0.00 -> L6 [6673996 6673998] (85MB), in 1.2s (1.2s total), output rate 71MB/s
```

```
[n2,s8,pebble] 302917  [JOB 19866] compacting(tombstone-density) L5 [6673335] (1.4MB) Score=0.00 + L6 [6670570 6668014 6668644 6668649 6673055 6673060 6673056 6672933 6672929 6667641 6672760 6668752 6668756 6668753 6668426 6670681 6670688 6670682 6669574 6669571 6669070 6672176 6672181 6672177 6668900 6670804 6670811 6670805 6664304 6665940 6665945 6669620 6669639 6672344 6672349 6672345 6669103 6669098 6667202 6671862 6671869 6671863 6671832 6671828 6667445 6667440 6664316 6669584 6669588 6669585 6668044 6670713 6670721 6672983 6673247 6673251 6673294 6673299 6673295 6673077 6673045 6673234 6673225 6673235 6673178 6673220 6673211 6673221 6673079 6673089 6673000 6673103 6673094 6673266 6673256 6673267 6673195 6673313 6673304 6673330 6673321 6673331 6673147 6673158 6673063 6673073 6672860 6672915 6672906 6673039 6673030 6673040 6672969 6673350 6673341 6673351 6673015 6673365 6673356 6673366 6673131 6673172 6673162 6673173 6673113 6673279 6673270 6673280 6673048 6673058 6672921 6672931 6672714 6672705 6672805 6672796 6673249 6673239 6673296 6673287 6673297 6672781 6672885 6672876 6672995 6672986 6672996 6672949 6672960 6672748 6672746 6672273 6673342 6673357 6672685 6673343 6673358 6673108 6673109] (266MB) Score=0.00; OverlappingRatio: Single 0.00, Multi 0.00
```

On top of that, we observed very high compaction counts in general, and excessively high compaction scores:

```
Store 2:
      |                             |       |       |   ingested   |     moved    |    written   |       |    amp
level | tables  size val-bl vtables | score |   in  | tables  size | tables  size | tables  size |  read |   r   w
------+-----------------------------+-------+-------+--------------+--------------+--------------+-------+---------
    0 |     0     0B     0B       0 |  0.00 | 195MB |   914  813KB |     0     0B |   898   41MB |  50KB |   0  0.2
    1 |     0     0B     0B       0 |  0.00 |    0B |     0     0B |     0     0B |     0     0B |    0B |   0  0.0
    2 |     0     0B     0B       0 |  0.00 |    0B |     0     0B |     0     0B |     0     0B |    0B |   0  0.0
    3 |     1  155KB  4.6KB       0 |  0.01 |  38MB |     3  2.7KB |    22  2.9MB |  1.5K  1.7GB | 1.7GB |   1 45.4
    4 |     2   34KB     0B       0 |  0.15 |  21MB |    12   14KB |     2   10MB |   113  385MB | 387MB |   1 18.1
    5 |   160  250MB  1.5MB      12 | 1317877001 |  29MB |   263  508KB |     0     0B |   195  1.3GB | 1.3GB |   1 45.3
    6 |  4.2K   63GB  232MB     888 |     - |  68MB |  4.3K   74MB |    54   52KB |  1.0K  4.8GB | 5.0GB |   1 72.1
total |  4.4K   64GB  233MB     900 |     - | 271MB |  5.5K   76MB |    78   13MB |  3.7K  8.4GB | 8.4GB |   4 31.9
-------------------------------------------------------------------------------------------------------------------
```

```
Store 3:
      |                             |       |       |   ingested   |     moved    |    written   |       |    amp
level | tables  size val-bl vtables | score |   in  | tables  size | tables  size | tables  size |  read |   r   w
------+-----------------------------+-------+-------+--------------+--------------+--------------+-------+---------
    0 |     0     0B     0B       0 |  0.00 | 228MB |    85   76KB |     0     0B |   102   34MB |    0B |   0  0.2
    1 |     0     0B     0B       0 |  0.00 |    0B |     0     0B |     0     0B |     0     0B |    0B |   0  0.0
    2 |     0     0B     0B       0 |  0.00 |    0B |     0     0B |     0     0B |     0     0B |    0B |   0  0.0
    3 |     0     0B     0B       0 |  0.00 |  20MB |     0     0B |    52   14MB |   145  404MB | 407MB |   0 20.3
    4 |     7  6.6KB     0B       0 |  0.14 |  59MB |     1   908B |    10   10MB |    77   82MB |  83MB |   1  1.4
    5 |    78   26MB  2.7MB       2 |  0.99 |  69MB |   180  369KB |     0     0B |  1.3K  539MB | 543MB |   1  7.8
    6 |  1.7K   49GB  251MB      58 |     - |  70MB |   304  314KB |     7  5.2KB |  1.7K   49GB |  50GB |   1  717
total |  1.8K   49GB  254MB      60 |     - | 230MB |   570  759KB |    69   25MB |  3.3K   50GB |  51GB |   3  223
-------------------------------------------------------------------------------------------------------------------
WAL: 1 files (1.7MB)  in: 175MB  written: 229MB (31% overhead) failover: (switches: 0, primary: 14m5.520801515s, secondary: 0s)
Flushes: 119
Compactions: 974  estimated debt: 0B  in progress: 1 (43MB)
             default: 98  delete: 21  elision: 251  move: 69  read: 0  tombstone-density: 535  rewrite: 0  copy: 0  multi-level: 0
MemTables: 1 (64MB)  zombie: 2 (128MB)
Zombie tables: 85 (975MB, local: 975MB)
Backing tables: 14 (399MB)
Virtual tables: 60 (225MB)
Local tables size: 49GB
Compression types: snappy: 1749 unknown: 60
Block cache: 219K entries (5.8GB)  hit rate: 82.4%
Table cache: 19K entries (14MB)  hit rate: 99.9%
Secondary cache: 0 entries (0B)  hit rate: 0.0%
Snapshots: 0  earliest seq num: 0
Table iters: 5
Filter utility: 66.9%
Ingestions: 95  as flushable: 30 (229KB in 180 tables)
Cgo memory usage: 6.5GB  block cache: 5.9GB (data: 5.8GB, maps: 21MB, entries: 17MB)  memtables: 640MB
```

Jira issue: PEBBLE-279

**Comment by jbowens** [comment]
Did we get to the bottom of where that excessive level score came from? The tombstone density compaction heuristic doesn't get incorporated into the level score, right?

**Comment by itsbilal** [comment]
@jbowens I made a change locally to panic if the range-annotated compensated size differed from the manually calculated compensated size, and it never triggered even after stressing the metamorphic tests. So I think that wasn't an issue. 

However a huge size is possible if the `PointDeletionsBytesEstimate` is significantly higher than the size of the level itself; like if the avg value size underneath each L5 file is pretty large. I'm convinced it's just something along these lines, and none of this code changed in 24.3 so it's likely a non-issue.

I then focused my attention on the very high number / size of tombstone-density compactions and devised #4074 to neuter them a bit.
